app/src/main/java/de/westnordost/streetcomplete/data/osm/elementgeometry/NodeWayMap.kt

```diff
@@ -3,21 +3,15 @@ package de.westnordost.streetcomplete.data.osm.elementgeometry
 
 /** Knows which vertices connect which ways. T is the identifier of a vertex  */
 class NodeWayMap<T>(ways: List<List<T>>) {
-    private val wayEndpoints = LinkedHashMap<T, MutableList<List<T>>>()
+    private val wayEndpoints: MutableMap<T, MutableList<List<T>>> = LinkedHashMap()
 
     init {
         for (way in ways) {
             val firstNode = way.first()
             val lastNode = way.last()
 
-            if (!wayEndpoints.containsKey(firstNode)) {
-                wayEndpoints[firstNode] = ArrayList()
-            }
-            if (!wayEndpoints.containsKey(lastNode)) {
-                wayEndpoints[lastNode] = ArrayList()
-            }
-            wayEndpoints[firstNode]!!.add(way)
-            wayEndpoints[lastNode]!!.add(way)
+            wayEndpoints.computeIfAbsent(firstNode, { ArrayList() }).add(way)
+            wayEndpoints.computeIfAbsent(lastNode, { ArrayList() }).add(way)
```

Comment:
That's nice! Though when I paste this line into my source code locally, Android Studio warns
`Call requires API level 24 (current min is 17): java.util.HashMap#computeIfAbsent`

Reply:
Yeah, that's because the implementation of `computeIfAbsent()` used in the desugaring process is included in the `Map` interface as a default method:

```
default V computeIfAbsent(K key,
            Function<? super K, ? extends V> mappingFunction) {
        Objects.requireNonNull(mappingFunction);
        V v;
        if ((v = get(key)) == null) {
            V newValue;
            if ((newValue = mappingFunction.apply(key)) != null) {
                put(key, newValue);
                return newValue;
            }
        }

        return v;
    }
```